src/daft-recordbatch/src/ops/window.rs

```diff
@@ -194,6 +350,250 @@ impl RecordBatch {
+                    (row_idx + 1) // End is exclusive, hence +1
+                        .saturating_add_signed(*offset as isize)
+                        .min(total_rows)
+                }
+                Some(WindowBoundary::RangeOffset(offset)) => {
+                    let offset = offset.to_series();
+                    let upper_bound = (current_row_order_by.clone() + offset)?;
+                    let gt = if descending {
+                        order_by_col.lt(&upper_bound)?
+                    } else {
+                        order_by_col.gt(&upper_bound)?
+                    };
+                    let mut first_true_idx = gt.len();
+                    for i in 0..gt.len() {
+                        if gt.get(i).unwrap_or(false) {
+                            first_true_idx = i;
+                            break;
+                        }
+                    }
+                    first_true_idx
+                }
+                Some(WindowBoundary::UnboundedPreceding)
+                | Some(WindowBoundary::UnboundedFollowing) => {
+                    unreachable!()
+                }
+            };
+
+            let Some(frame_size) = frame_end.checked_sub(frame_start) else {
+                return Err(DaftError::ValueError(
+                    "Negative frame size is not allowed".into(),
+                ));
+            };
+
+            // Check min_periods requirement
+            if frame_size >= min_periods {
+                // Remove values that left the window (values that were in the previous window but not in the current one)
+                if frame_start > prev_frame_start {
+                    agg_state.remove(prev_frame_start, frame_start)?;
+                }
+
+                // Add new values that entered the window (values that are in the current window but weren't in the previous)
+                if frame_end > prev_frame_end {
+                    agg_state.add(prev_frame_end, frame_end)?;
+                }
+
+                // Update previous boundaries for the next iteration
+                prev_frame_start = frame_start;
+                prev_frame_end = frame_end;
+                validity.push(true);
+            } else {
+                validity.push(false);
+            }
+
+            // Evaluate current state to get the result for this row
+            agg_state.evaluate()?;
+        }
+
+        let mut validity = Bitmap::from(validity);
+        let agg_state = agg_state.build()?.rename(name);
+        if let Some(agg_validity) = agg_state.validity() {
+            validity = binary(&validity, agg_validity, |a, b| a & b);
+        }
+
+        // Build the final result series
+        let renamed_result = agg_state.with_validity(Some(validity))?;
+        let window_batch = Self::from_nonempty_columns(vec![renamed_result])?;
+        self.union(&window_batch)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn window_agg_range(
```

Comment:
I'm seeing a lot of similar code between `window_agg_range_incremental` and `window_agg_range`. Maybe we can build some helper functions? I'm not sure